### PR TITLE
Fix the usage of linear variables in handlers

### PR DESCRIPTION
**Pull request by thwfhk**
This PR fixes the issue #1161 as well as some other problems of using linear variables in handlers:

1. Fix the usage counting of variables from environment in handler clauses;
2. Check the usage times of linear parameters of handler clauses;
3. Prohibit the usage of linear variables from environment in deep handlers.

Also, it moves the definition of module `Timestamp` in `commonTypes.ml` to a more reasonable position.

### CODE REVIEW

File: core/commonTypes.ml [hunk 1]
```diff
@@ -42,78 +42,6 @@ module DeclaredLinearity = struct
     | _   -> false
 end
 
-module Timestamp = struct
```
Comment (dhil) [inline core/commonTypes.ml:45]: Preferably done in a separate PR. But no problems this time.

File: core/typeSugar.ml [hunk 2]
```diff
@@ -1789,6 +1789,9 @@ end = struct
           'sink'. *)
        empty
     | usages :: usagess  ->
+       (* let () = StringMap.iter (fun s _ -> print_string <| s ^ "  ") usages in
+       let _ = List.map (StringMap.iter (fun s _ -> print_string <| s ^ "  ")) usagess in
+       print_string "\n" ; *)
```
Comment (dhil) [inline core/typeSugar.ml:1794]: Delete?

Comment (dhil) [inline core/typeSugar.ml:1794]:
```suggestion
```

File: core/typeSugar.ml [hunk 3]
```diff
@@ -2262,6 +2265,7 @@ let type_pattern ?(linear_vars=true) closed
     then Types.fresh_type_variable (lin_any, res_any)
     else Types.fresh_type_variable (lin_unl, res_any) in
 
+  (* NOTE: outdated comments? basic idea is still correct *)
```
Comment (dhil) [inline core/typeSugar.ml:2268]: I suppose we should update the comment to reflect what the code is **really** doing...

File: core/typeSugar.ml [hunk 4]
```diff
@@ -4139,11 +4147,39 @@ let rec type_check : context -> phrase -> phrase * Types.datatype * Usage.t =
...
+                        if uses <> 1 then
+                          if Types.Unl.can_type_be t then
+                            Types.Unl.make_type t
+                          else
+                            Gripers.non_linearity pos uses v t)
+                      (pattern_env pat)
+                   in
+                   (* check the usages of environment linear variables in deep handlers *)
+                   let () =
+                    if descr.shd_depth = Deep then
```
Comment (dhil) [inline core/typeSugar.ml:4172]: What about shallow handlers?

File: core/typeSugar.ml [hunk 5]
```diff
@@ -4139,11 +4147,39 @@ let rec type_check : context -> phrase -> phrase * Types.datatype * Usage.t =
...
+                      Usage.iter
+                        (fun v _ ->
+                          if not (StringSet.mem v vs) then
+                            let t = Env.find v henv.var_env in
+                            if Types.Unl.can_type_be t then
+                              Types.Unl.make_type t
+                            else
+                              Gripers.die pos ("Variable " ^ v ^ " of linear type " ^ Types.string_of_datatype t ^ " is used in a deep handler."))
```
Comment (dhil) [inline core/typeSugar.ml:4180]: You can define a `Griper` for this message. See the `Gripers` code near the top of the file.

File: core/typeSugar.ml [hunk 6]
```diff
@@ -4156,12 +4192,44 @@ let rec type_check : context -> phrase -> phrase * Types.datatype * Usage.t =
...
+                   (* check the usages of linear parameters in handler clauses *)
+                   let () =
+                    Env.iter (fun v t ->
+                      let uses = Usage.uses_of v (usages body) in
+                        if uses <> 1 then
+                          if Types.Unl.can_type_be t then
+                            Types.Unl.make_type t
+                          else
+                            Gripers.non_linearity pos uses v t)
+                      (pattern_env pat)
```
Comment (dhil) [inline core/typeSugar.ml:4216]: Same check as above. Can we abstract this?

File: tests/handler_linearity.tests [hunk 7]
```diff
@@ -0,0 +1,42 @@
+Use linear variables in deep handlers (1)
```
Comment (dhil) [inline tests/handler_linearity.tests:1]: Excellent! The tests all seem to be "failing tests", i.e. they are meant to fail. It would be good to add some "success tests", i.e. tests that are meant to succeed.

**Review by dhil — CHANGES_REQUESTED** [review]
LGTM modulo change requests.

File: core/typeSugar.ml (hunk 3 above)
Comment (thwfhk) [inline core/typeSugar.ml:2265]:
```suggestion
```
Probably it is still meaningful 🤔. Just delete my comments.

File: core/typeSugar.ml (hunk 4 above)
Comment (thwfhk) [reply core/typeSugar.ml:4168]: Shallow handlers can use linear variables from environments.

File: tests/handler_linearity.tests (hunk 7 above)
Comment (thwfhk) [reply tests/handler_linearity.tests:1]: Not so easy to add success tests. I added one test about using linear vars in clauses of shallow handlers. Not sure how to write operations which use linear parameters 🤔

**Comment by thwfhk** [comment]
Probably ready to merge now?

File: tests/handlers/lin_deep1.links [hunk 8]
```diff
@@ -0,0 +1,11 @@
+var lf = linfun(x) {x};
+
+handle({
+  ignore(do Foo);
+  lf(2)
+}) {
+  case Foo(resume) -> {
```
Comment (dhil) [inline tests/handlers/lin_deep1.links:7]:
```suggestion
  case <Foo => resume> -> {
```

File: tests/handlers/lin_deep2.links [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+});
+
+handle({
+  ignore(do Foo);
+  var (i, ch) = receive(ch);
+  println("Int: " ^^ intToString(i));
+  close(ch)
+}) {
+  case Return(_) -> ()
+  case Flip(resume) -> {
```
Comment (dhil) [inline tests/handlers/lin_deep2.links:13]:
```suggestion
  case _ -> ()
  case <Flip => resume> -> {
```

File: tests/handlers/lin_deep3.links [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+  var ch = fork(fun(ch) {
+     var ch = send(42, ch);
+     close(ch)
+  });
+
+  handle({
+    ignore(do Flip(ch))
+  }) {
+    case Return(_) -> ()
+    case Flip(ch, resume) -> {
```
Comment (dhil) [inline tests/handlers/lin_deep3.links:11]:
```suggestion
    case _ -> ()
    case <Flip(ch) => resume> -> {
```

File: tests/handlers/lin_deep4.links [hunk 11]
```diff
@@ -0,0 +1,11 @@
+var lf = linfun(x) {x};
+
+handle({
+  ignore(do Foo);
+  lf(2)
+}) {
+  case Return(_) -> lf(3)
+  case Foo(resume) -> {
```
Comment (dhil) [inline tests/handlers/lin_deep4.links:8]:
```suggestion
  case _ -> lf(3)
  case <Foo => resume> -> {
```

File: tests/handlers/lin_shallow1.links [hunk 12]
```diff
@@ -0,0 +1,12 @@
+{
+  var lf = linfun(x) {x};
+
+  shallowhandle({
+    ignore(do Foo);
+    lf(1)
+  }) {
+    case <Foo => _> -> {
```
Comment (dhil) [inline tests/handlers/lin_shallow1.links:8]:
```suggestion
    case <Foo -> _> -> {
```

File: tests/handlers/lin_shallow2.links [hunk 13]
```diff
@@ -0,0 +1,11 @@
+{
+  var lf = linfun(x) {x};
+
+  shallowhandle({
+    ignore(do Foo);
+    lf(1)
+  }) {
+    case _ -> lf(2)
+    case <Foo => _> -> 3
```
Comment (dhil) [inline tests/handlers/lin_shallow2.links:9]:
```suggestion
    case <Foo -> _> -> 3
```

File: tests/handlers/lin_shallow3.links [hunk 14]
```diff
@@ -0,0 +1,12 @@
+{
+  var lf = linfun(x) {x};
+
+  shallowhandle({
+    ignore(do Foo);
+    1
+  }) {
+    case <Foo => _> -> {
```
Comment (dhil) [inline tests/handlers/lin_shallow3.links:8]:
```suggestion
    case <Foo -> _> -> {
```

**Review by dhil — CHANGES_REQUESTED** [review]
Some of the tests are using the old syntax for operation patterns.

**Review by dhil — APPROVED** [review]
